packages/Search/src/details/DTK_DetailsTreeTraversal.hpp

```diff
@@ -181,21 +181,43 @@ KOKKOS_FUNCTION int nearestQuery( BoundingVolumeHierarchy<DeviceType> const bvh,
         double const node_distance = queue.top().second;
         // NOTE: it would be nice to be able to do something like
         // tie( node, node_distance = queue.top();
-        queue.pop();
+
         if ( TreeTraversal<DeviceType>::isLeaf( bvh, node ) )
         {
+            queue.pop();
             insert( TreeTraversal<DeviceType>::getIndex( bvh, node ),
                     node_distance );
             count++;
         }
         else
         {
-            // insert children of the node in the priority list
-            for ( Node const *child :
-                  {node->children.first, node->children.second} )
+            auto const c1 = node->children.first;
+            auto const c2 = node->children.second;
+
+            auto const c1d = distance( c1 );
+            auto const c2d = distance( c2 );
+
+            if ( c1 != nullptr && c2 != nullptr )
+            {
+
+                if ( c1d > c2d )
+                {
+                    queue.replace_top( c1, c1d );
+                    queue.push( c2, c2d );
+                }
+                else
+                {
+                    queue.replace_top( c2, c2d );
+                    queue.push( c1, c1d );
+                }
+            }
+            else if ( c1 != nullptr )
+            {
+                queue.replace_top( c1, c1d );
```

Comment:
No you've got to ensure that `( c1d <= queue.top().second )` before you call `replace_top`.